### PR TITLE
Add initial script generator DSL

### PR DESCRIPTION
**Pull request by Octogonapus**
### Description of the Change

This PR adds the foundation of our code generation logic. The generated scripts are described in a Gradle-inspired DSL. This description is then transformed into a directed code dependency graph and turned into code.

### Motivation

This is the foundation of the backend :)

### Verification Process

Unit tests for all relevant new features. Integration testing for the DSL. 

Manual verification of the generated output from running the script :horse: :
image

### Applicable Issues

None.

### CODE REVIEW

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/Code.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package edu.wpi.axon.dsl
+
+import edu.wpi.axon.dsl.variable.Variable
+
+/**
+ * A "typeclass" for generating code. The code this interface means is (possibly multiline)
+ * computations that can assign to variables and can depends on other variables or computations.
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/Code.kt]: Typo `depends`

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import arrow.data.fix
+import arrow.data.k
+import com.google.common.graph.GraphBuilder
+import com.google.common.graph.ImmutableGraph
+import com.google.common.graph.MutableGraph
+import edu.wpi.axon.dsl.container.PolymorphicNamedDomainObjectContainer
+
+/**
+ * Parses a [PolymorphicNamedDomainObjectContainer] of [Code] into an [ImmutableGraph], assuring
+ * that all [Code] nodes are unique and dependencies are properly ordered.
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt]: The uniqueness claim should be amended with something like "nodes on separate branches of the graph will not be deduplicated".

```
+---+      +---+
| A +----->+ B +--------+
+---+      +---+        v
                      +-+-+
                      | D |
                      +-+-+
+---+      +---+        ^
| A +----->+ C +--------+
+---+      +---+
```

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ */
+@Suppress("UnstableApiUsage")
+class CodeGraph(
+    private val container: PolymorphicNamedDomainObjectContainer<Code<Code<*>>>
+) {
+
+    val graph: ImmutableGraph<Code<Code<*>>> by lazy {
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt]: Add a typealias for `Code<Code<*>>`.

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    private fun populateGraphUsingDependencies(
+        graph: MutableGraph<Code<Code<*>>>,
+        code: Code<Code<*>>
+    ) {
+        // Make sure nodes without dependencies are in the graph
+        graph.addNode(code)
+
+        code.dependencies.forEach {
+            // Don't recursive if the edge was already in the graph
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt]: Typo `recursive`

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    ): Boolean {
+        graph.successors(root).forEach { node ->
+            val reachable = bfs(graph, node)
+
+            if (visited anyIn reachable) {
+                return true
+            }
+
+            if (hasCircuits(graph, node, visited + node)) {
+                return true
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/CodeGraph.kt:107]: Clean up this logic.

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/PolymorphicNamedDomainObjectContainerUtil.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import edu.wpi.axon.dsl.task.Task
+import edu.wpi.axon.dsl.variable.Variable
+import kotlin.reflect.KClass
+
+/**
+ * Creates a new variable and configures it.
+ */
+fun <T : Variable, U : T> PolymorphicNamedDomainObjectContainer<T>.creating(
+    type: KClass<U>,
+    configuration: (U.() -> Unit)? = null // TODO: Remove configuration for variables
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/PolymorphicNamedDomainObjectContainerUtil.kt]: This should be resolved before merging.

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/ScriptGenerator.kt [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+class ScriptGenerator(
+    val variables: PolymorphicNamedDomainObjectContainer<Variable>,
+    val tasks: PolymorphicNamedDomainObjectContainer<Task>,
+    configure: ScriptGenerator.() -> Unit
+) : Configurable {
+
+    // TODO: Do something with this (probably make the script implement a method and return this)
+    // var scriptOutput: Variable? = null
+
+    // TODO: Find an intelligent way to derive this instead of needing it to be specified
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/ScriptGenerator.kt:25]: Once `CodeGraph` verifies there are no islands, we should be able to start from any node and find the last node in the DAG.

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/ScriptGenerator.kt [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    fun code(generateDebugComments: Boolean = false) = buildString {
+        require(isConfiguredCorrectly()) {
+            "The DSl was not configured correctly."
+        }
+
+        (variables.values + tasks.values + requiredVariables)
+            .filter { !it.isConfiguredCorrectly() }
+            .let {
+                if (it.isNotEmpty()) {
+                    throw IllegalArgumentException(
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/ScriptGenerator.kt]: Refactor this into `require(it.isEmpty())`.

File: core/dsl/src/main/kotlin/edu/wpi/axon/dsl/task/ConstructYoloV3ImageInput.kt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Constructs the input argument for the YoloV3 model.
+ */
+class ConstructYoloV3ImageInput(name: String) : Task(name) {
+
+    var imageDataInput: Variable? = null
+    var imageSizeInput: Variable? = null
+    var sessionInput: Variable? = null
+    var output: Variable? = null
```
Comment (Octogonapus) [inline core/dsl/src/main/kotlin/edu/wpi/axon/dsl/task/ConstructYoloV3ImageInput.kt]: This pattern in all the Tasks is really gross. I'm sort of remembering a `singleAssign` delegate, maybe in Koin? Logically, these variables are `lateinit`, but you can't check whether a `lateinit var` has been initialized the same way you can check a nullable variable isn't `null`.

Comment (Octogonapus) [reply core/dsl/src/main/kotlin/edu/wpi/axon/dsl/task/ConstructYoloV3ImageInput.kt]: I was thinking of TornadoFX's `singleAssign`. https://github.com/edvin/tornadofx/blob/1eec3eb5f694cf55d9f3c567130df873167d9b88/src/main/java/tornadofx/Properties.kt#L141

**Review by AustinShalit — COMMENTED** [review]
How do you feel about putting all of the yolo specific files into their own directory?

**Comment by Octogonapus** [comment]
> How do you feel about putting all of the yolo specific files into their own directory?

I have moved the yolov3-specific Tasks into their own package.

**Comment by Octogonapus** [comment]
@AustinShalit How is the documentation? Do I need to elaborate on anything?

**Review by AustinShalit — APPROVED** [review]

**Pull request merged**